File: src/main/java/net/netty/PacketDecoder.java
package net.netty;

import io.netty.buffer.ByteBuf;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.ByteToMessageDecoder;
import util.crypto.MapleAESOFB;
import util.crypto.ShandaCrypto;
import util.packet.Packet;

import java.util.List;

public class PacketDecoder extends ByteToMessageDecoder {

    @Override
    protected void decode(ChannelHandlerContext chc, ByteBuf in, List<Object> out) {
        NettyClient c = chc.channel().attr(NettyClient.Companion.getCLIENT_KEY()).get();
        MapleAESOFB mCr = chc.channel().attr(NettyClient.Companion.getCRYPTO_KEY()).get();

        if (c != null) {
            byte[] iv = c.getRecvIV();
            if (c.getStoredLength() == -1) {
                if (in.readableBytes() >= 4) {
                    int h = in.readInt();
                    if (!MapleAESOFB.checkPacket(h, iv)) {
                        c.close();
                        return;
                    }
                    c.setStoredLength(MapleAESOFB.getLength(h));
                } else {
                    return;
                }
            }
            if (in.readableBytes() >= c.getStoredLength()) {
                byte[] dec = new byte[c.getStoredLength()];
                in.readBytes(dec);
                c.setStoredLength(-1);

                dec = mCr.crypt(dec, iv);
                c.setRecvIV(MapleAESOFB.getNewIv(iv));

                ShandaCrypto.decrypt(dec);

                out.add(new Packet(dec));
            }
        }
    }
}

File: src/main/java/net/netty/PacketDecoder.kt
package net.netty

import io.netty.buffer.ByteBuf
import io.netty.channel.ChannelHandlerContext
import io.netty.handler.codec.ByteToMessageDecoder
import net.netty.NettyClient.Companion.CLIENT_KEY
import net.netty.NettyClient.Companion.CRYPTO_KEY
import util.crypto.MapleAESOFB
import util.crypto.ShandaCrypto
import util.packet.Packet

class PacketDecoder : ByteToMessageDecoder() {

    override fun decode(chc: ChannelHandlerContext, byteBuf: ByteBuf, out: MutableList<Any>) {
        val c = chc.channel().attr(CLIENT_KEY).get()
        val mCr = chc.channel().attr(CRYPTO_KEY).get()

        if (c != null) {
            val iv = c.recvIV
            if (c.storedLength == -1) {
                if (byteBuf.readableBytes() >= 4) {
                    val h = byteBuf.readInt()
                    if (!MapleAESOFB.checkPacket(h, iv)) {
                        c.close()
                        return
                    }
                    c.storedLength = MapleAESOFB.getLength(h)
                } else {
                    return
                }
            }
            if (byteBuf.readableBytes() >= c.storedLength) {
                var dec = ByteArray(c.storedLength)
                byteBuf.readBytes(dec)
                c.storedLength = -1

                dec = mCr.crypt(dec, iv)
                c.recvIV = MapleAESOFB.getNewIv(iv)

                ShandaCrypto.decrypt(dec)

                out.add(Packet(dec))
            }
        }
    }
}